<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Table of contents</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

</style>
        
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="table-of-contents">Table of contents</h1>
<ul>
<li><a href="#table-of-contents">Table of contents</a></li>
<li><a href="#starting">Starting</a></li>
<li><a href="#configuring-snakemake">Configuring Snakemake</a>
<ul>
<li><a href="#input-dataset">Input dataset</a></li>
<li><a href="#assembly">Assembly</a></li>
<li><a href="#deseq2-model">Deseq2 Model</a></li>
<li><a href="#transposable-elements-te">Transposable Elements (TE)</a></li>
<li><a href="#single-sample">Single Sample</a>
<ul>
<li><a href="#genes">Genes</a></li>
<li><a href="#tes">TEs</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#running-snakemake">Running Snakemake</a></li>
<li><a href="#output">Output</a></li>
<li><a href="#troubleshooting">Troubleshooting</a>
<ul>
<li><a href="#fixing-read-names-for-optical-duplicates">Fixing read names for optical duplicates</a></li>
<li><a href="#pipeline-not-starting">Pipeline not starting</a></li>
<li><a href="#some-step-in-the-pipeline-fails">Some step in the pipeline fails</a></li>
</ul>
</li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
</ul>
<h1 id="starting">Starting</h1>
<p>This readme provides instructions on how to run the snakemake RNA-seq pipeline on a cluster.</p>
<p>To begin you will need to grab a copy of the pipeline from the shared folder in the cluster:</p>
<blockquote>
<p>hpc2:<br>
/home/share/dcyleung/snakemake/RNA-seq/<br>
biocrfhpc1:<br>
/home4/share/dcyleung/snakemake/RNA-seq/<br>
biocrfhpc2:<br>
/data1/share/dcyleung/Pipeline/snakemake/RNA-seq/</p>
</blockquote>
<p>and place it in your own folder, preferentially in its own folder since snakemake will create many new files.</p>
<p>You will see 2 folders, one named workflow (which includes the code and scripts, please <em><strong>do not</strong></em> modify anything in this folder) and a config folder; and the run_snk.sh script that will be used to send the pipeline to slurm.</p>
<p>The config folder includes the following files:</p>
<ul>
<li>config.yaml</li>
<li>samples.tsv</li>
<li>units.tsv</li>
</ul>
<p>Due to the difference in the folder structure in the clusters, the config file of the pipeline is configured for a specific cluster and cannot be used with the other one. Specifically, config.yaml and run_snk.sh are specific to the cluster, while samples.tsv, units.tsv and the files in workflow folder can be used in any cluster.</p>
<p>Before starting, make sure that your fastq files are compressed in fastq.gz format, both to save space in clusters and to make sure the pipeline works properly.</p>
<h1 id="configuring-snakemake">Configuring Snakemake</h1>
<h2 id="input-dataset">Input dataset</h2>
<p>The previously mentioned <em>samples.tsv</em> has to be modified according to your own samples. It is recommended to modify the tsv files in Excel or similar programs since a misplaced tabulation will cause snakemake to improperly reading the config file and probably stop it from working. If you use Excel, please <strong>make sure you save the file as a csv or tsv file</strong> and not as an Excel file, otherwise the pipeline cannot read the file.</p>
<p>The first file you need to modify is <em>samples.tsv</em>. It is a tab separated file that looks like this:</p>
<blockquote>
<table>
<thead>
<tr>
<th>sample_name</th>
<th>condition</th>
</tr>
</thead>
<tbody>
<tr>
<td>A1</td>
<td>treated</td>
</tr>
<tr>
<td>B1</td>
<td>untreated</td>
</tr>
<tr>
<td>A2</td>
<td>treated</td>
</tr>
<tr>
<td>B2</td>
<td>untreated</td>
</tr>
</tbody>
</table>
</blockquote>
<p>You need to modify this file to include any samples you want to analyze in the pipeline, and the condition that will be used in Deseq2 model. If you are not going to use Deseq2 leave condition as <em>treated</em> for every sample, but the samples should still be filled in this file.</p>
<p>If you need to analyze more conditions in Deseq2, you can add more columns to the right of <em>condition</em> column and set each sample treatment in that column.</p>
<p>The next file that needs to be modified is <em>units.tsv</em>, where you indicate the location of your fastq.gz files. The unit_name columns refer to technical replicates of a sample, e.g. lanes in sequencing. Different units from the same sample will be merged together (eg. Sample A1 lane1 and lane2 fq files will be merged). If your samples don't have technical replicates, give the same unit_name to all samples. This file looks like this:</p>
<blockquote>
<table>
<thead>
<tr>
<th>sample_name</th>
<th>unit_name</th>
<th>fq1</th>
<th>fq2</th>
<th>sra</th>
<th>strandedness</th>
</tr>
</thead>
<tbody>
<tr>
<td>A1</td>
<td>lane1</td>
<td>A1.lane1.R1.fastq.gz</td>
<td>A1.lane1.R2.fastq.gz</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>A1</td>
<td>lane2</td>
<td>A1.lane2.R1.fastq.gz</td>
<td>A1.lane2.R2.fastq.gz</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>B1</td>
<td>lane1</td>
<td>B1.lane1.R1...</td>
<td>B1.lane1.R2...</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>B1</td>
<td>lane2</td>
<td>...</td>
<td>...</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>A2</td>
<td>lane1</td>
<td>...</td>
<td>...</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>A2</td>
<td>lane2</td>
<td>...</td>
<td>...</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>B2</td>
<td>lane1</td>
<td>...</td>
<td>...</td>
<td></td>
<td>1</td>
</tr>
<tr>
<td>B2</td>
<td>lane2</td>
<td>...</td>
<td>...</td>
<td></td>
<td>1</td>
</tr>
</tbody>
</table>
</blockquote>
<p>You will need to fill this file with either the location of your fastq.gz files or an sra ID for public samples. The path to your files can be the full path to your files, i.e:</p>
<blockquote>
<p>/root/user/snakemake/samples/sample_1/sample_1_R1.fastq.gz</p>
</blockquote>
<p>or a relative path from where snakemake is run, which is the directory where the folders <em>workflow</em> and <em>config</em> lie, i.e:</p>
<blockquote>
<p>samples/sample_1/sample_1_R1.fastq.gz</p>
</blockquote>
<p>This last approach is the preferred one.</p>
<p><strong>If only the column fq1 is filled, snakemake will run the pipeline as single end. If both fq1 and fq2 are filled, snakemake will run the pipeline as paired end. Whether SRA accession samples are considered paired or single end is determined by the <em>single_end</em> setting activation in <em>config.yaml</em>. This setting must still be set appropriately for local samples, as a few steps in the pipeline require it. If both SRA and fastq.gz are present, snakemake will use the fastq.</strong></p>
<p>The last necessary column is strandedness, which <strong>needs to be equal for all units</strong>. The number equivalences can be found here:</p>
<blockquote>
<table>
<thead>
<tr>
<th>Number</th>
<th>Strandedness</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>Stranded</td>
</tr>
<tr>
<td>0.5</td>
<td>Unstranded</td>
</tr>
<tr>
<td>0</td>
<td>Reverse Stranded*</td>
</tr>
<tr>
<td>*Only used in Illumina TruSeq protocol</td>
<td></td>
</tr>
</tbody>
</table>
</blockquote>
<p>Lastly, the <em>config.yaml</em> file sets what analyses snakemake will do. This file has been commented to briefly explain what each setting does, so modify the settings to your needs.</p>
<p>However, some options require further explanation that can be seen bellow.</p>
<h2 id="assembly">Assembly</h2>
<p>To indicate which genome assembly you want to use for the reference genome you have to modify the assembly field in <em>config.yaml</em>. Currently the pipeline supports UCSC and Genecode assemblies. To use a Genecode assembly you have to provide the Genecode reference to the assembly field.</p>
<p>Four Genecode assemblies are supported:</p>
<blockquote>
<table>
<thead>
<tr>
<th>Organism</th>
<th>assembly</th>
<th>Ucsc rmsk</th>
</tr>
</thead>
<tbody>
<tr>
<td>Human</td>
<td>GRCh38</td>
<td>hg38</td>
</tr>
<tr>
<td>Human</td>
<td>GRCh37</td>
<td>hg19</td>
</tr>
<tr>
<td>Mouse</td>
<td>GRCm39</td>
<td>mm39</td>
</tr>
<tr>
<td>Mouse</td>
<td>GRCm38</td>
<td>mm10</td>
</tr>
</tbody>
</table>
</blockquote>
<p>In case you are using the pipeline for TEs, the remasker file indicated in the 3rd column will be downloaded from UCSC to be used as a reference.</p>
<p>If you indicate any assembly name that does not appear in the table, the pipeline will try to download an UCSC reference genome. For example, if you want to use the reference mouse genome from UCSC, you can indicate mm39 instead of GCRm39 in the assembly field.</p>
<p>For organisms other than human or mouse simply indicate the assembly name present in UCSC browser.</p>
<h2 id="deseq2-model">Deseq2 Model</h2>
<p>To use Deseq2 there needs to be a model set in <em>config.yaml</em>. The model needs to be set even if you are not going to activate Deseq2, in this case just leave the default model as is. The default model used is</p>
<pre><code>~condition
</code></pre>
<p>which uses the condition column present in <em>samples.tsv</em>. If you added additional columns to be used for testing in Deseq2, they need to be put in the model. For example, if you added the column <em>mutated</em> to <em>samples.tsv</em>, the model present in <em>config.yaml</em> would be:</p>
<pre><code>~condition + mutated
</code></pre>
<p>Then Deseq2 will produce a result file by comparing 2 contrasts, which are indicated in <em>config.yaml</em> contrasts section. A base one is already provided in <em>config.yaml</em>, but more can be added in the following way:</p>
<pre><code>contrasts:
    treated-vs-untreated:
      - treated
      - untreated
    mutated-vs-control:
      - mutated
      - control    
</code></pre>
<p>where mutated-vs-control is what will be used for the file name, and the conditions inside is the contrasts that will be compared in Deseq2. Note that all the contrasts here have to appear in <em>samples.tsv</em> for Deseq2 to work.</p>
<h2 id="transposable-elements-te">Transposable Elements (TE)</h2>
<p>Within the DEseq2 config block is also the TE configuration. If you activate the TE option snakemake will produce DEseq2 output for TEs in addition to the normal DEseq2 output using the RepeatMasker file from UCSC for your selected assembly, including the R object and the pca (if activated). Both DEseq2 and TEs will use the same model specified in this config block.</p>
<p>The other setting in this section is the filter, which will filter out elements in DEseq2 with less than the filter number of counts. Is set to 10 by default but you can change it to suit your needs.</p>
<h2 id="single-sample">Single Sample</h2>
<h3 id="genes">Genes</h3>
<p>To be used when you have no replicates for a sample. It is activated by turning the single option in config.yaml to True, and it will perform an alternative analysis to Deseq2. Therefore, the single option cannot be used at the same time at Deseq2 (and therefore pca), otherwise the pipeline will not work. Single mode also requires mergeReads to be activated, as it will be comparing samples and not units.</p>
<p>Additionally, the treatment of one of the samples in samples.tsv <strong>must be called <code>control</code></strong>. The other treatments do not matter, as every sample will be compared to the control sample.</p>
<p>Once you run the pipeline in single mode with one sample as control, you can rerun it with another sample as control and new files comparing samples to the new control will be generated without needing to run the whole pipeline.</p>
<h3 id="tes">TEs</h3>
<p>In a similar way, you can obtain TE differential expression for samples with no replicates by turning <em>TE_single</em> to activate in the <code>config.yaml</code> file. TE_single shares the same requirements as single mode does, and both single and TE_single can be activated at the same time, producing output for both genes and TEs.</p>
<p>The only difference for TE_single is the filter, which will remove from the analysis elements with less than the filter number of counts, similarly to the TE analysis with DEseq2.</p>
<h1 id="running-snakemake">Running Snakemake</h1>
<p>To run the pipeline simply send the run_snk.sh script to slurm by doing:</p>
<pre><code>sbatch run_snk.sh
</code></pre>
<p>from the directory where the folders you copied, config and workflow, are located. If run_snk.sh is run from another directory snakemake will not detect the pipeline and it will not work. This will run snakemake with the default config for sbatch for the cluster, using the partition general for hpc2 and q1 for biocrf clusters.</p>
<p>Inside <em>run_snk.sh</em> you can modify the partition to which snakemake will run by changing the line <code>#SBATCH -p q1</code>, which is only one job. As for the partition where snakemake will send the jobs, you can change the last 2 lines of the <code>run_snk.sh</code> file. If you want to use the another partition, comment (using #) the snakemake line with the profile and uncomment the longer snakemake line (by removing #). Then set your partition by changing or adding <code>-p &lt;partition&gt;</code> after the <code>sbatch</code> command present in the line (within the single quotes).</p>
<p>Additionally, snakemake is configured to run up to 10 simultaneous jobs by default in biocrfhpcs, 2 for hpc2; but if you're using the pipeline in a cluster with number of jobs restrictions you may want to reduce it to 2 or 1 jobs by changing the <code>-j</code> parameter in the long <code>snakemake</code> line. Snakemake will take at least 2 job spots, one for the pipeline manager and another for the actual jobs, when <code>-j</code> is set to 1. In hpc2, <code>-j</code> is set to 2 by default since its the cluster with the most restrictions. Therefore if you use the pipeline in the biocrfhpc clusters it will be faster.</p>
<h1 id="output">Output</h1>
<p>Snakemake will store all the output files in a directory called results. The outputs are organized by steps in the pipeline and samples, so you will se folders like star, rsem, etc; with sample folders within them. For example:</p>
<pre><code>results/rsem/pe/sample1/star_output_files
</code></pre>
<p>In results, the qc folder will contain the files <code>multiqc_report_data</code> and <code>multiqc_report.html</code>, which includes fastqc and rseqc. You can download these files and view in a browser.</p>
<p>If you activated deseq2, in the deseq2 results folder you can find <code>all.rds</code> which contains a Deseq object called dds of all your data. You can load it in R in case you want to further explore the data with Deseq2. The pipeline will also produce some plots like the pca (if activated) so you can initially asses your data.</p>
<p>For single and TE_single modes, the pipeline will create the single and TE_single folders in results respectively, where files named {sample}_vs_{control}.tsv will be placed, comparing all the samples in <code>samples.tsv</code> against the sample marked as control. These files contain the differential expression between each sample and the control. You can rerun the pipeline with a different sample marked as control, which will run only the last step to compare samples against a new control and therefore will be much quicker.</p>
<p>Additionally logs for each step will be stored in the logs folder.</p>
<h1 id="troubleshooting">Troubleshooting</h1>
<h2 id="fixing-read-names-for-optical-duplicates">Fixing read names for optical duplicates</h2>
<p>In order for picard to pick up reads that are duplicated in close positions, read names need to include coordinates from the sequencer.</p>
<p>These coordinates look like this (last part of the name):</p>
<blockquote>
<p>VL00299:10:AAC522NM5:1:1101:27415:1133</p>
</blockquote>
<p>So reads that come from our lab should already be ok, however for public data SRA changes the naming format to the following:</p>
<blockquote>
<p>@SRR25205388.1 VL00299:10:AAC522NM5:1:1101:27415:1133 length=76</p>
</blockquote>
<p>As STAR removes anything from the read name after the first whitespace, you need to fix your reads to remove the first whitespace so they look like this:</p>
<blockquote>
<p>@SRR25205388.1_VL00299:10:AAC522NM5:1:1101:27415:1133 length=76</p>
</blockquote>
<p>SRA reads downloaded from the pipeline are automatically corrected, however if you downloaded your reads on your own you need to fix their name. To do this you can use the following command:</p>
<blockquote>
<p>Fastq.gz: <br>
zcat <Your fastq.gz> | sed -e 's/^(@[^[:blank:]]<em>)[[:blank:]]+/\1_/' -e 's/^(+[^[:blank:]]</em>)[[:blank:]]+/\1_/' | gzip &gt; &lt;Fixed fastq.gz&gt;</p>
<p>Fastq: <br>
cat &lt;Your fastq&gt; | sed -e 's/^(@[^[:blank:]]<em>)[[:blank:]]+/\1_/' -e 's/^(+[^[:blank:]]</em>)[[:blank:]]+/\1_/' &gt; &lt;Fixed fastq&gt;</p>
</blockquote>
<p>If the fastq file is missing these coordinates on the read names then optical duplicates cannot be removed.
The picard report for duplicate detection is included in <code>report.html</code> file, so you can check if optical duplicates were correctly removed or not.</p>
<h2 id="pipeline-not-starting">Pipeline not starting</h2>
<p>Most of these issues have to do with the formatting of units.tsv or samples.tsv, but you may have also activated incompatible config settings.</p>
<ul>
<li>'utf-8' codec can't decode...</li>
</ul>
<p>If this line appears in the error message in the pipeline slurm output file, it means that there is a special character somewhere in either <code>units.tsv</code> or <code>samples.tsv</code> (usually the message specifies which file). Libreoffice's calc and nano can show special characters, so you can use them to check your files and remove them.</p>
<p>Another possibility is that you saved any of the tsv files as an Excel .xml file, in which case they won't work unless you save them as csv files with tab separators.</p>
<ul>
<li>Wrong samples or units formatting</li>
</ul>
<p>If snakemake complains that it cannot find a column in one of the tsv files, or that the values in the column are the wrong type of data; it usually means that there are wrongly placed tabs in the file. You may want to check that the columns are properly formatted in a spreadsheet program and also try to check if there are extra tabs anywhere in the file (like another line at the bottom, or after the last column).</p>
<ul>
<li>Directory locked</li>
</ul>
<p>If snakemake complains that the directory is locked, it means that either snakemake is already running or that the previous snakemake job crashed or was cancelled and could not unlock the directory. If snakemake is still running wait until it finishes. If the directory is locked and snakemake is not running, run the following code to unlock the directory (from the RNA-seq directory):</p>
<pre><code> source ~/../share/dcyleung/miniconda3/etc/profile.d/conda.sh 
 conda activate snakemake 
 snakemake --unlock 
 conda deactivate 
 source ~/.bashrc
</code></pre>
<h2 id="some-step-in-the-pipeline-fails">Some step in the pipeline fails</h2>
<p>The only step that I've found that can fail that is not my fault (ehem) is rsem. There seems to be a bug where rsem is failing to calculate the confidence intervals (ci), seems to happen more with sra samples. If this is the case, deactivate ci for rsem in <code>config.yaml</code>, there are further instructions in the file.</p>
<p>Another step that can fail with a not useful log is Deseq2 init (thanks R), which usually has to do with problems with the model. You can try to run with only <code>~condition</code> as a model to check that it works.</p>
<p>You should also double check that the <code>single</code> setting in the config is set correctly, as some steps in the pipe will fail if set for paired end when reading single end files.</p>
<p>If any other step is failing for you, please tell me so I can fix it.</p>
<h1 id="acknowledgements">Acknowledgements</h1>
<p>This pipeline is based on the pipeline made by jafors:</p>
<p><a href="https://doi.org/10.5281/zenodo.5245549"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5245549.svg" alt="DOI"></a> <a href="https://github.com/snakemake-workflows/rna-seq-star-deseq2">Snakemake workflow: rna-seq-star-deseq2</a></p>

        
        
    </body>
    </html>