<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Table of contents</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="table-of-contents">Table of contents</h1>
<ul>
<li><a href="#table-of-contents">Table of contents</a></li>
<li><a href="#starting">Starting</a></li>
<li><a href="#configuring-snakemake">Configuring Snakemake</a>
<ul>
<li><a href="#input-dataset">Input dataset</a></li>
<li><a href="#mode-of-use">Mode of use</a></li>
<li><a href="#restriction-enzymes">Restriction Enzymes</a></li>
<li><a href="#assembly">Assembly</a></li>
</ul>
</li>
<li><a href="#running-snakemake">Running Snakemake</a></li>
<li><a href="#output">Output</a></li>
<li><a href="#too-long-dont-want-to-read">Too long, don't want to read</a></li>
<li><a href="#references">References</a></li>
</ul>
<h1 id="starting">Starting</h1>
<p>This readme provides instructions on how to run the snakemake Hi-C pipeline on a cluster.</p>
<p>To begin you will need to grab a copy of the pipeline from the shared folder in the cluster:</p>
<blockquote>
<p>hpc2:<br>
/home/share/dcyleung/snakemake/Hi-C/<br>
biocrfhpc1:<br>
/home4/share/dcyleung/snakemake/Hi-C/<br>
biocrfhpc2:<br>
/data1/share/dcyleung/Pipeline/snakemake/Hi-C/</p>
</blockquote>
<p>and place it in your own folder, preferentially in its own folder since snakemake will create many new files.</p>
<p>You will see 2 folders, one named workflow (which includes the code and scripts, please <em><strong>do not</strong></em> modify anything in this folder) and a config folder; and the run_snk.sh script that will be used to send the pipeline to slurm.</p>
<p>The config folder includes the following files:</p>
<ul>
<li>config.yaml</li>
<li>samples.tsv</li>
<li>units.tsv</li>
</ul>
<p>Due to the difference in the folder structure in the clusters, the config file of the pipeline is configured for a specific cluster and cannot be used with the other one. Particularly, config.yaml and run_snk.sh are specific to the cluster, while samples.tsv, units.tsv and the files in workflow folder can be used in any cluster.</p>
<h1 id="configuring-snakemake">Configuring Snakemake</h1>
<h2 id="input-dataset">Input dataset</h2>
<p>The previously mentioned <em>samples.tsv</em> has to be modified according to your own samples. It is recommended to modify the tsv files in Excel or similar programs since a misplaced tabulation will cause snakemake to improperly reading the config file and probably stop it from working. If you use Excel, please <strong>make sure you save the file as a csv or tsv file</strong> and not as an Excel file, otherwise the pipeline cannot read the file.</p>
<p>The first file you need to modify is <em>samples.tsv</em>. It is a tab separated file that looks like this:</p>
<blockquote>
<table>
<thead>
<tr>
<th>sample_name</th>
<th>group</th>
<th>batch_effect</th>
</tr>
</thead>
<tbody>
<tr>
<td>J1_Va_Rep1</td>
<td>J1</td>
<td>batch1</td>
</tr>
<tr>
<td>J1_Va_Rep2</td>
<td>J1</td>
<td>batch1</td>
</tr>
<tr>
<td>J1_Va_Rep3</td>
<td>J1</td>
<td>batch1</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep1</td>
<td>DnmtTKO</td>
<td>batch2</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep2</td>
<td>DnmtTKO</td>
<td>batch2</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep3</td>
<td>DnmtTKO</td>
<td>batch2</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep1</td>
<td>DnmtTKO</td>
<td>batch3</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep1</td>
<td>DnmtTKO</td>
<td>batch3</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep1</td>
<td>DnmtTKO</td>
<td>batch3</td>
</tr>
</tbody>
</table>
</blockquote>
<p>You need to modify this file to include any samples you want to analyze in the pipeline, along with their group, which represents biological replicates of the same sample. <strong>Samples belonging to the same group can be merged to create a Hi-C matrix for all the biological replicates together.</strong> The batch effect column currently does nothing, but it is good to keep track of it.</p>
<p><strong>It is also advisable to avoid special characters (like - or |) in the name of the samples as some of them are used by the pipeline to process results, but the pipeline should still work with them.</strong></p>
<p>The next file that needs to be modified is <em>units.tsv</em>, where you indicate the location of your fastq.gz files. The unit column refer to technical replicates of a sample, e.g. lanes in sequencing. This file looks like this:</p>
<blockquote>
<table>
<thead>
<tr>
<th>sample_name</th>
<th>unit</th>
<th>fq1</th>
<th>fq2</th>
<th>sra</th>
<th>platform</th>
</tr>
</thead>
<tbody>
<tr>
<td>J1_Va_Rep1</td>
<td>1</td>
<td>J1_1.lane1.R1.fastq.gz</td>
<td>J1_1.lane1.R2.fastq.gz</td>
<td></td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>J1_Va_Rep2</td>
<td>1</td>
<td>J1_2.lane1.R1.fastq.gz</td>
<td>J1_2.lane1.R2.fastq.gz</td>
<td></td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>J1_Va_Rep3</td>
<td>1</td>
<td>J1_3.lane1.R1.fastq.gz</td>
<td>J1_3.lane1.R2.fastq.gz</td>
<td></td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep1</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194959</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep2</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194960</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Co_Rep3</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194961</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep1</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194962</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep2</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194963</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>DnmtTKO_Va_Rep3</td>
<td>1</td>
<td></td>
<td></td>
<td>SRR10194964</td>
<td>ILLUMINA</td>
</tr>
</tbody>
</table>
</blockquote>
<p>You will need to fill this file with either the location of your fastq.gz files or an sra ID for public samples. The path to your files can be the full path to your files, i.e:</p>
<blockquote>
<p>/root/user/snakemake/samples/sample_1/sample_1_R1.fastq.gz</p>
</blockquote>
<p>or a relative path from where snakemake is run, which is the directory where the folders <em>workflow</em> and <em>config</em> lie, i.e:</p>
<blockquote>
<p>samples/sample_1/sample_1_R1.fastq.gz</p>
</blockquote>
<p>This latter approach is the preferred one.</p>
<p><strong>Note that this pipeline currently only supports Paired End samples as most methods currently available only accept Paired End samples. Running it will Single End samples will probably crash the pipeline</strong></p>
<p>Lastly, the <em>config.yaml</em> file sets what analyses snakemake will do. This file has been commented to explain what each setting does, so modify the settings to your needs.</p>
<p>Hi-C many different parameters for each step, each one of them is briefly explained in the config.yaml file itself. As this pipeline uses <a href="https://github.com/vaquerizaslab/fanc">fanC</a>[1] as a base, it is recommended that you also check their <a href="https://fan-c.readthedocs.io/en/latest/">documentation</a></p>
<h2 id="mode-of-use">Mode of use</h2>
<p>This pipeline can be used in different ways. The main &quot;switches&quot; are found in the fanc section of config.yaml: <code>merge_groups</code> at the start of the fanc section and <code>activate</code> and <code>pca_only</code> in the analysis section.</p>
<ul>
<li>
<p>If you only want to get the <code>.hic</code> files and do your own analysis and plots, turn off the analysis activation so the pipeline only maps your fastqs and generates the matrix.</p>
</li>
<li>
<p>If you want to analyze your data is advised that you first check how similar your biological replicates are in order to be merged safely. To do this, run the pipeline with merge_groups as <code>False</code> and pca_only as <code>True</code> with the analysis activated. In this case, each sample will have its own Hi-C matrix generated and compared by doing a PCA plot.</p>
</li>
<li>
<p>If you are confident in your biological replicates or the PCA looks ok, you can then run the analysis with pca_only as <code>False</code> and merge_groups as <code>True</code>, which will produce the rest of the analysis.</p>
</li>
</ul>
<p><strong>Do note that Hi-C can be really slow if you use the full genome, so if you are interested only in specific chromosomes or regions, it is better to restrict the pipeline to only map to those regions in the <code>chr</code> option of the fanc section.</strong></p>
<h2 id="restriction-enzymes">Restriction Enzymes</h2>
<p>In Hi-C, the genome needs to be &quot;digested&quot; and separated into fragments that change depending on the restriction enzymes utilized by the protocol
used to sequence your samples. The pipeline can digest the genome if you provide the name of the restriction enzymes as long as they're found in <a href="http://rebase.neb.com/rebase/rebase.html">REBASE</a>. By default the pipeline is set to digest using the restriction enzymes used by the common Arima protocol.</p>
<h2 id="assembly">Assembly</h2>
<p>To indicate which genome assembly you want to use for the reference genome you have to modify the assembly field in <em>config.yaml</em>. Currently the pipeline supports UCSC and Genecode assemblies. To use a Genecode assembly you have to provide the Genecode reference to the assembly field.</p>
<p>Four Genecode assemblies are supported:</p>
<blockquote>
<table>
<thead>
<tr>
<th>Organism</th>
<th>assembly</th>
</tr>
</thead>
<tbody>
<tr>
<td>Human</td>
<td>GRCh38</td>
</tr>
<tr>
<td>Human</td>
<td>GRCh37</td>
</tr>
<tr>
<td>Mouse</td>
<td>GRCm39</td>
</tr>
<tr>
<td>Mouse</td>
<td>GRCm38</td>
</tr>
</tbody>
</table>
</blockquote>
<p>If you indicate any assembly name that does not appear in the table, the pipeline will try to download an UCSC reference genome. For example, if you want to use the reference mouse genome from UCSC, you can indicate mm39 instead of GCRm39 in the assembly field.</p>
<p>For organisms other than human or mouse simply indicate the assembly name present in UCSC browser.</p>
<h1 id="running-snakemake">Running Snakemake</h1>
<p>To run the pipeline simply send the run_snk.sh script to slurm by doing:</p>
<pre><code>sbatch run_snk.sh
</code></pre>
<p>from the directory where the folders you copied, config and workflow, are located. If run_snk.sh is run from another directory snakemake will not detect the pipeline and it will not work.</p>
<p>Inside <em>run_snk.sh</em> you can modify the partition to which snakemake will send the jobs by changing the line <code>#SBATCH -p q1</code> and the option <code>-p q1</code> in the <code>snakemake</code> line if its present (in hpc2 and biocrfhpc1). snakemake will run on a default profile settings in every cluster, if you want to use the another partition, comment (using #) the snakemake line with the profile and uncomment the longer snakemake line (by removing #). Then set your partition by changing or adding <code>-p &lt;partition&gt;</code> after the <code>sbatch</code> command present in the line (within the single quotes).</p>
<p>Additionally, snakemake is configured to run up to 5 simultaneous jobs by default, but if you're using the pipeline in a cluster with number of jobs restrictions you may want to reduce it to 2 or 1 jobs by changing the <code>-j</code> parameter in the <code>snakemake</code> line. Snakemake will take at least 2 job spots, one for the pipeline manager and another for the actual jobs, when <code>-j</code> is set to 1. In hpc2, <code>-j</code> is set to 1 by default since its the cluster with the most restrictions. Therefore if you use the pipeline in the biocrfhpc clusters it will be faster.</p>
<h1 id="output">Output</h1>
<p>Snakemake will store all the output files in a directory called results. The outputs are organized by steps in the pipeline and samples, so you will se folders like bwa, fastqc, etc; with sample folders within them. For example:</p>
<pre><code>results/bwa/sample1_bwa_outputs
</code></pre>
<p>In results, the qc folder will contain the files <code>multiqc_report_data</code> and <code>multiqc_report.html</code>, which includes fastqc and rseqc, along with the folder ATACseqQC, where all its plots are found. You can download these files and view in a browser. The file <code>report.html</code> will also be found in the root directory of the pipeline, in which you can find the rest of the QC and stats that are not included in multiqc.</p>
<p>In addition to the plots and qc that will be present in these files, the pipeline will provide you with 3 Hi-C matrices files:</p>
<ul>
<li>A FanC Hi-C file that can be used with FanC software to make plots.</li>
<li>A Juicer Hi-C file with multiple resolutions that can be used with the juicebox software tools.</li>
<li>A cooler Hi-C file that can be used with cooler and domaincaller.</li>
</ul>
<p>Additionally logs for each step will be stored in the logs folder.</p>
<h1 id="too-long-dont-want-to-read">Too long, don't want to read</h1>
<p>The basic steps to run the pipe are:</p>
<ul>
<li>Make a copy of the pipeline.</li>
<li>Put your samples and groups in samples.tsv.</li>
<li>Put your units and file paths/sra codes in units.tsv</li>
<li>Change config.yaml to your restriction enzymes, set assembly and any specific step option you need.</li>
<li>Use <code>sbatch</code> to send run_snk.sh to the cluster.</li>
</ul>
<h1 id="references">References</h1>
<p>[1] Kruse, K., Hug, C.B. &amp; Vaquerizas, J.M.
FAN-C: a feature-rich framework for the analysis and visualisation of chromosome conformation capture data.
Genome Biol 21, 303 (2020).
<a href="https://doi.org/10.1186/s13059-020-02215-9">https://doi.org/10.1186/s13059-020-02215-9</a></p>

        
        
    </body>
    </html>